angel_system/global_step_prediction/global_step_predictor.py

```diff
@@ -0,0 +1,596 @@
+                    self.trackers[tracker_ind]["prediction_history"], current_step
+                )
+        # Update the current_frame
+        self.current_frame += len(activity_confs)
+
+        self.activity_conf_history = np.append(
+            self.activity_conf_history, activity_confs, axis=0
+        )
+
+        return self.trackers
+
+    def plot_gt_vs_predicted_one_recipe(self, step_gts, fname_suffix=None):
+        # Plot gt vs predicted class across all vid frames
+        fig = plt.figure()
+        sn.set(font_scale=1)
+        step_gts = [float(i) for i in step_gts]
+        plt.plot(step_gts, label="gt")
+        for i, tracker in enumerate(self.trackers):
+            step_predictions = tracker["prediction_history"]
+            plt.plot(step_predictions, label=f"estimated_{i}")
+
+        starting_zero_value = 0
+        for i in range(len(self.avg_probs)):
+            starting_zero_value -= 2
+            plot_line = np.asarray(self.activity_conf_history)[:, i]
+            plt.plot(2 * plot_line + starting_zero_value, label=f"act_preds[{i}]")
+            plt.plot(
+                [
+                    2 * int(j < self.avg_probs[i]) + starting_zero_value
+                    for j in plot_line
+                ],
+                label=f"act_preds_threshold[{i}]",
+            )
+        plt.legend()
+        if not fname_suffix:
+            fname_suffix = f"vid{vid_id}"
+        fig.savefig(f"./outputs/plot_pred_vs_gt_{fname_suffix}.png")
+        # plt.show()
+
+    def sanitize_str(self, str_: str):
+        """
+        Convert string to lowercase and emove trailing whitespace and period.
+
+        :param str_: Input text
+
+        :return: ``str_`` converted to lowercase and stripped of trailing whitespace and period.
+        :rtype: str
+        """
+        return str_.lower().strip(" .")
+
+    def manually_increment_current_step(self, tracker_index):
+        self.trackers[tracker_index]["current_step"] += 1
+        return self.trackers
+
+    def manually_decrement_current_step(self, tracker_index):
+        self.trackers[tracker_index]["current_step"] -= 1
+        return self.trackers
+
+
+def plot_positive_GT_conf_distributions(activity_confs, activity_gt):
+    """
+    plot_TP_conf_distributions:
+    For each activity, plot the distribution of confidences when ground
+    truth indicates that activity is happening.
+
+    i.e.: for activity x, for frames in which ground truth = x, plot
+    the distribution of confidences.
+
+    Inputs:
+    activity_confs: frames x class-wise-confidences. Given a kwcoco
+        dataset called "coco":
+        ```
+        activity_confs = torch.asarray(coco.images().lookup("activity_conf"))
+        ```
+        (49K x 25 for coffee val set.)
+    activity_gt: frames x ground truth activity_id.
+        Given a kwcoco dataset called "coco":
+        ```
+        activity_gt = torch.asarray(coco.images().lookup("activity_gt"))
+        ```
```

Comment:
Very nearly the Google style doc-string!
```suggestion
    Args:
        activity_confs: frames x class-wise-confidences. Given a kwcoco
            dataset called "coco":
            ```
            activity_confs = torch.asarray(coco.images().lookup("activity_conf"))
            ```
            (49K x 25 for coffee val set.)
        activity_gt: frames x ground truth activity_id.
            Given a kwcoco dataset called "coco":
            ```
            activity_gt = torch.asarray(coco.images().lookup("activity_gt"))
            ```
```